optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bi/AbstractBiJoiner.java

```diff
@@ -75,11 +77,14 @@ public boolean matches(A a, B b) {
 
     public abstract Function<A, Object> getLeftMapping(int index);
 
-    public abstract Function<A, Object[]> getLeftCombinedMapping();
-
-    public abstract Function<B, Object> getRightMapping(int index);
-
-    public abstract Function<B, Object[]> getRightCombinedMapping();
+    public Function<A, Object[]> getLeftCombinedMapping() {
+        Function<A, Object>[] mappings = IntStream.range(0, getJoinerTypes().length)
+                .mapToObj(this::getLeftMapping)
+                .toArray(Function[]::new);
+        return (A a) -> Arrays.stream(mappings)
+                .map(f -> f.apply(a))
```

Comment:
You can probably use f::apply (one less wrapper lambda)

Reply:
Not in this case. It would work if I was trying to get rid of `f`, but here I need to pass `a` to `f` and therefore I need to do it like this.